sound/soc/sof/loader.c

```diff
@@ -358,6 +360,75 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
 	return ext_man_size;
 }
 
+static int snd_sof_fw_ext_man4_parse(struct snd_sof_dev *sdev, const struct firmware *fw)
+{
+	struct sof_ipc4_fw_modules *module_entry;
+	struct sof_man4_fw_binary_header *fw_header;
+	struct sof_ext_manifest4_hdr *hdr;
+	struct sof_man4_module_config *fm_config;
+	struct sof_man4_module_entry *fm_entry;
+	int i;
+
+	if (fw->size < sizeof(hdr)) {
+		dev_err(sdev->dev, "Invalid fw size\n");
+		return -EINVAL;
+	}
+
+	hdr = (struct sof_ext_manifest4_hdr *)fw->data;
+
+	fw_header = (struct sof_man4_fw_binary_header *)(fw->data + hdr->len +
+							 SOF_MAN4_FW_HDR_OFFSET);
+	dev_dbg(sdev->dev, " fw %s: header length %x, module num %d", fw_header->name,
+		fw_header->len, fw_header->num_module_entries);
+
+	fm_entry = (struct sof_man4_module_entry *)((void *)fw_header + fw_header->len);
+	fm_config = (struct sof_man4_module_config *)(fm_entry + fw_header->num_module_entries);
+
+	sdev->fw_modules = devm_kmalloc_array(sdev->dev, fw_header->num_module_entries,
+					      sizeof(*module_entry), GFP_KERNEL);
+	if (!sdev->fw_modules)
+		return -ENOMEM;
+
+	sdev->fw_module_num = fw_header->num_module_entries;
+	module_entry = sdev->fw_modules;
+
+	for (i = 0; i < fw_header->num_module_entries; i++) {
+		dev_dbg(sdev->dev, "module %s : UUID %pUL, ", fm_entry->name,
+			fm_entry->uuid);
+
+		memcpy(&module_entry->man4_module_entry, fm_entry, sizeof(*fm_entry));
+
+		if (fm_entry->cfg_count)
+			module_entry->bss_size = fm_config[fm_entry->cfg_offset].is_bytes;
+
+		module_entry->man4_module_entry.id = i;
+
+		/* align to dw */
+		module_entry->instance_id =
```

Comment:
what is the purpose of the instance ID? It doesnt look like it is incremental per module.

Reply:
> what is the purpose of the instance ID? It doesnt look like it is incremental per module.

The `instance_id` is to map modules and SOF widget. A module will map to a SOF widget and they will have the same id. @RanderWang Please correct me if I was wrong.

Reply:
> Ok 2 questions then,.why is it an array? Can we drop it for now and add it later when you show how it is used?

Sorry, "The instance_id is to map modules and SOF widget. A module will map to a SOF widget and they will have the same id." is not precise. `instance_id` here is a bit field. Each bit represents an active swidget and the amount of swidget (`instance_max_count`) is dynamic. `instance_max_count` could be 5 or 50 for example. That's why we use n dynamic size array. We could use `u64 instance_id;` if `instance_max_count` is guarantee less than 64. 
Yes, I can drop it for now and add it later.